draft-rha-jose-hpke-encrypt.md

```diff
@@ -199,6 +199,14 @@ The recipient will create the receiving HPKE context by invoking SetupBaseR() (S
 
 The Open function will, if successful, decrypts "ct".  When decrypted, the result will be either the CEK (when Key Encryption mode is used), or the content (if Integrated Encryption mode is used).  The CEK is the symmetric key used to decrypt the ciphertext.
 
+## Protected Header Parameters
+
+In order to ensure the security properties of HPKE, the following requirements MUST be followed for protected header parameters:
+
+* For Integrated Encryption, "epk" MUST be present, and MUST contain an Encapsulated JSON Web Key.
+* For Integrated Encryption, "alg" MUST be "dir", and "enc" MUST be an HPKE algorithm from JSON Web Signature and Encryption Algorithms in {{JOSE-IANA}}.
+* For Key Encryption, "alg" MUST NOT be present and "enc" MUST be a content encryption algorithm from JSON Web Signature and Encryption Algorithms in {{JOSE-IANA}}.
```

Comment:
> For Key Encryption, "alg" MUST NOT be present and "enc" MUST be a content encryption algorithm from JSON Web Signature and Encryption Algorithms in {{JOSE-IANA}}.

How? wouldn't alg be something like "HPKE...KE" (KW for Key Wrapping, KE for Key encryption?)

Reply:
Or the same as when used in "enc" but i'm not sure how that falls in line with the JOSE convention.

Reply:
In your implementation:

Single recipient JWE Protected header looks like this:

```
{
      enc: 'A128GCM',
      epk: {
        x: 'nu8aBst8LHUtXKNofo7EpTNalUfAL-k5cLuR_aGYXe8',
        crv: 'P-256',
        kty: 'EC',
        y: '3a8bD0CVfAYuuM_zv5KNiQPr82OU3MLU_thZ0yrGHSo'
      }
    }
```

Multiple recipient JWE Protected header looks like this:

```
{ enc: 'A128GCM' }
```

Neither has alg, hence in our "Key Encryption mode (starting think its a bad name)".... 

Our JWE Protected header should look like:

single recipient:

```
{
      enc: 'A128GCM',
      epk: {
        kty: 'EK',
        ek: '3a8bD0CVfAYuu...alUfAL-k5cLuR_aGYXe8'
      }
    }
```

multiple recipient:
```
{ enc: 'A128GCM' }
```

Reply:
> In your implementation:

A JWE JOSE Header (that's the combination of the three locations which are required to have disjoint properties) always has an `alg`.

Reply:
Ah, this whole section is only about "Protected Header Parameters". Sorry. Merge it, i'll make a suggestion over the merged text. 

This whole paragraph should be dispersed among the modes' actual definitions and saying that "alg" MUST NOT be present is just incorrect. A mode which supports multiple recipients doesn't necessarily mean that it is exclusively used for multiple recipients. Hence "alg" may actually be present when the mode is used for a single recipient and it may then be protected. But even when used with multiple recipients, if all recipients use the same "alg", which would be common in closed ecosystems with dictated algorithm support, there's no reason why the "alg" couldn't be protected then. It's best to not mention it altogether because this protected/unprotected comes implied from the mode/number of recipients.

The only requirements that should be present are for when something MUST be integrity protected; therefore, it MUST occur only within the JWE Protected Header.